cpp/src/arrow/array/array_primitive.h

```diff
@@ -90,26 +90,25 @@ class NumericArray : public PrimitiveArray {
   using value_type = typename TypeClass::c_type;
   using IteratorType = stl::ArrayIterator<NumericArray<TYPE>>;
 
-  explicit NumericArray(const std::shared_ptr<ArrayData>& data) : PrimitiveArray(data) {}
+  explicit NumericArray(const std::shared_ptr<ArrayData>& data) { SetData(data); }
 
   // Only enable this constructor without a type argument for types without additional
   // metadata
   template <typename T1 = TYPE>
   NumericArray(enable_if_parameter_free<T1, int64_t> length,
                const std::shared_ptr<Buffer>& data,
                const std::shared_ptr<Buffer>& null_bitmap = NULLPTR,
-               int64_t null_count = kUnknownNullCount, int64_t offset = 0)
-      : PrimitiveArray(TypeTraits<T1>::type_singleton(), length, data, null_bitmap,
-                       null_count, offset) {}
-
-  const value_type* raw_values() const {
-    return reinterpret_cast<const value_type*>(raw_values_) + data_->offset;
+               int64_t null_count = kUnknownNullCount, int64_t offset = 0) {
+    SetData(ArrayData::Make(TypeTraits<T1>::type_singleton(), length, {null_bitmap, data},
```

Comment:
I think this is a breaking change (for TimestampArray), where PrimitiveArray::SetData is called in the PrimitiveArray::PrimitiveArray. The code below can reproduce this issue.

```
TEST(TestPrettyPrintArray, TimestampArray) {
  auto array = std::make_shared<TimestampArray>(
      timestamp(TimeUnit::MICRO),
      /*length=*/1, Buffer::FromVector(std::vector<int64_t>{0}));
  auto v0 = array->GetView(0);
  std::cout << v0 << std::endl;
}
```

cc @mapleFU @pitrou @raulcd

Reply:
I'm integrating Arrow 18.0.0 to our internal code base and this break the CI. Not sure if this is a blocker to RC0.

Reply:
I've check here and this is because 

```
template <typename TYPE>
class NumericArray : public PrimitiveArray {
  using PrimitiveArray::PrimitiveArray;
};
```

This call `PrimitiveArray` without setting `values_` in  `NumericArray`, leaving a uninitialized pointer here...

Reply:
I've created an issue here: https://github.com/apache/arrow/issues/44541